game/fs.cc

```diff
@@ -32,46 +32,218 @@ namespace {
 	}
 
 	const fs::path performous = "performous";
-	const fs::path configSchema = "config/schema.xml";	
-
-	struct PathCache {
-		Paths paths;
-		bool didMigrateConfig = false;
-		fs::path base, share, locale, sysConf, home, conf, data, cache;
-		/// Expand a path specifier as a list of actual paths. Expands ~ (home) and DATADIR (Performous search path).
-	Paths pathExpand(fs::path p) {
-			Paths ret;
-			if (pathRootHack(p, "~")) ret.push_back(home / p);
-			else if (pathRootHack(p, "DATADIR")) {
-				// Add all data paths with p appended to them
-				for (auto const& path: paths) ret.push_back(path / p);
+	const fs::path configSchema = "config/schema.xml";
+
+	std::mutex mutex;
+	using Lock = std::lock_guard<std::mutex>;
+
+	void PathCache::pathBootstrap() {
+		if (!base.empty()) return;  // Only bootstrap once
+		// Base (e.g. /usr/local), share (src or installed data files) and locale (built or installed .mo files)
+		{
+			char const* root = getenv("PERFORMOUS_ROOT");
+			switch (Platform::currentOS()) {
+				case Platform::HostOS::OS_WIN:
+					base = fs::canonical(root ? root : execname().parent_path());
+					break;
+				case Platform::HostOS::OS_MAC:
+					if (const auto appFolder{execname().parent_path().parent_path().parent_path()}; appFolder.extension() == ".app") {
+						base = fs::canonical(appFolder);
+						break;
+					}
+					[[fallthrough]]; // if not running from .app bundle, handle like regular unix.
+				default:
+					base = fs::canonical(root ? root : execname().parent_path().parent_path());
+					break;
+			}
+			std::set<fs::path> infixPaths{fs::path(SHARED_DATA_DIR), fs::path("data"),fs::path()};
+			do {
+				if (base.empty()) throw std::runtime_error("Unable to find Performous data files. Install properly or set environment variable PERFORMOUS_ROOT.");
+				for (auto const& infix : infixPaths) {
+					if (!fs::exists(base / infix / configSchema)) continue;
+					share = base / infix;
+					break; // Found
+				}
+				if (!share.empty()) {
+					break; // Found
+				}
+				base = base.parent_path();
+			} while (true);
+			// Use locale .mo files from build folder?
+			if (Platform::currentOS() == Platform::HostOS::OS_WIN) {
+				auto folder = base.filename();
+				auto reg = "x\\d{2}-([Dd]ebug|[Rr]elease)(-install)?"; // matches all build folders.
```

Comment:
We're missing one case; DebInfo

and maybe a couple more when running on windows but through wsl

Reply:
Would you mind adding those, since that's more your area?